Gifski/SavePanelAccessoryViewController.swift

```diff
@@ -1,71 +1,174 @@
+
 	@IBOutlet private var estimatedSizeLabel: NSTextField!
-	@IBOutlet private var scaleSlider: NSSlider!
-	@IBOutlet private var scaleLabel: NSTextField!
 	@IBOutlet private var frameRateSlider: NSSlider!
 	@IBOutlet private var frameRateLabel: NSTextField!
 	@IBOutlet private var qualitySlider: NSSlider!
+
+	@IBOutlet private var widthTextField: NSTextField!
+	@IBOutlet private var heightTextField: NSTextField!
+	@IBOutlet private var predefinedSizesDropdown: NSPopUpButton!
+	@IBOutlet private var dimensionsTypeDropdown: NSPopUpButton!
+
 	var inputUrl: URL!
 	var videoMetadata: AVURLAsset.VideoMetadata!
 	var onDimensionChange: ((CGSize) -> Void)?
 	var onFramerateChange: ((Int) -> Void)?
 
+	let formatter = ByteCountFormatter()
+
+	private var resizableDimensions: ResizableDimensions!
+	private var predefinedSizes: [PredefinedSizeItem]!
+
 	override func viewDidLoad() {
 		super.viewDidLoad()
 
-		let formatter = ByteCountFormatter()
 		formatter.zeroPadsFractionDigits = true
+		setupDimensions()
+		setupDropdowns()
+		setupSliders()
+	}
+
+	private func setupDimensions() {
+		let minimumScale: CGFloat = 0.01
+		let maximumScale: CGFloat = 1.0
+		let dimensions = Dimensions(type: .pixels, value: videoMetadata.dimensions)
+		resizableDimensions = ResizableDimensions(dimensions: dimensions, minimumScale: minimumScale, maximumScale: maximumScale)
+
+		let pixelCommonSizes: [CGFloat] = [dimensions.value.width, 960.0, 800.0, 640.0, 500.0, 480.0, 320.0, 256.0, 200.0, 160.0, 128.0, 80.0, 64.0].sorted(by: >)
+		let pixelDimensions = pixelCommonSizes.map { width -> CGSize in
+			let ratio = width / dimensions.value.width
+			let height = dimensions.value.height * ratio
+			return CGSize(width: width, height: height)
+		}
+		let predefinedPixelDimensions = pixelDimensions
+			.filter { resizableDimensions.validate(newSize: $0) }
+			.map { resizableDimensions.resized(to: $0) }
 
-		// TODO: Use KVO here
+		let percentCommonSizes: [CGFloat] = [100.0, 50.0, 33.0, 25.0, 20.0]
+		let predefinedPercentDimensions = percentCommonSizes
+			.map {
+				resizableDimensions.changed(dimensionsType: .percent)
+					.resized(to: CGSize(width: $0, height: $0))
+			}
 
-		var currentDimensions = videoMetadata.dimensions
+		predefinedSizes = [.custom]
+		predefinedSizes.append(.spacer)
+		predefinedSizes.append(contentsOf: predefinedPixelDimensions.map { .dimensions($0) })
+		predefinedSizes.append(.spacer)
+		predefinedSizes.append(contentsOf: predefinedPercentDimensions.map { .dimensions($0) })
+	}
 
-		func estimateFileSize() {
-			let frameCount = videoMetadata.duration * frameRateSlider.doubleValue
-			var fileSize = (Double(currentDimensions.width) * Double(currentDimensions.height) * frameCount) / 3
-			fileSize = fileSize * (qualitySlider.doubleValue + 1.5) / 2.5
-			estimatedSizeLabel.stringValue = formatter.string(fromByteCount: Int64(fileSize))
+	private func setupDropdowns() {
+		predefinedSizesDropdown.removeAllItems()
+		predefinedSizes.forEach { size in
```

Comment:
I prefer using normal for loops.

Reply:
Oh sure, I don't have a strong opinion on that one so will replace it 👍